# Handling duplicate, missing, or invalid data
## 중복, 결측, 유효하지 않은 데이터 다루기

## About the data
In this notebook, we will using daily weather data that was taken from the [National Centers for Environmental Information (NCEI) API](https://www.ncdc.noaa.gov/cdo-web/webservices/v2) and altered to introduce many common problems faced when working with data. 

*Note: The NCEI is part of the National Oceanic and Atmospheric Administration (NOAA) and, as you can see from the URL for the API, this resource was created when the NCEI was called the NCDC. Should the URL for this resource change in the future, you can search for "NCEI weather API" to find the updated one.*

## Background on the data

Data meanings:
- `PRCP`: precipitation in millimeters / 밀리미터 단위의 강수량
- `SNOW`: snowfall in millimeters / 밀리미터 단위의 강설량
- `SNWD`: snow depth in millimeters / 밀리미터 단위의 적설량
- `TMAX`: maximum daily temperature in Celsius / 일일 최고 섭씨 온도
- `TMIN`: minimum daily temperature in Celsius / 일일 최저 섭씨 온도
- `TOBS`: temperature at time of observation in Celsius / 관측 시의 섭씨 온도
- `WESF`: water equivalent of snow in millimeters / 밀리미터 단위의 눈에 해당하는 물의 양.

Some important facts to get our bearings:
- According to the National Weather Service, the coldest temperature ever recorded in Central Park was -15°F (-26.1°C) on February 9, 1934: [source](https://www.weather.gov/media/okx/Climate/CentralPark/extremes.pdf) 
- The temperature of the Sun's photosphere is approximately 5,505°C: [source](https://en.wikipedia.org/wiki/Sun)

## Setup

In [3]:
import pandas as pd

df = pd.read_csv('data/dirty_data.csv')

## Finding problematic data
### 데이터셋에서 문제가 있는 데이터 찾기

**1. row 행들 살펴보기**

In [4]:
df.head() 

date            station  PRCP  SNOW  SNWD    TMAX  TMIN  \
0  2018-01-01T00:00:00                  ?   0.0   0.0  -inf  5505.0 -40.0   
1  2018-01-01T00:00:00                  ?   0.0   0.0  -inf  5505.0 -40.0   
2  2018-01-01T00:00:00                  ?   0.0   0.0  -inf  5505.0 -40.0   
3  2018-01-02T00:00:00  GHCND:USC00280907   0.0   0.0  -inf    -8.3 -16.1   
4  2018-01-03T00:00:00  GHCND:USC00280907   0.0   0.0  -inf    -4.4 -13.9   

   TOBS  WESF inclement_weather  
0   NaN   NaN               NaN  
1   NaN   NaN               NaN  
2   NaN   NaN               NaN  
3 -12.2   NaN             False  
4 -13.3   NaN             False

-> 
* station 필드에는 ?와 관측소 ID가 섞여 있다.
* SNWD 적설량에는 음의 무한대값(-inf)가 있다.
* TMAX 최고 섭씨 온도에는 너무 큰 값이 있다.(ex. 5505.0)
* inclement_weather 필드와 몇몇 열에는 많은 NaN 값이 있다.

**2. describe() 요약통계를 사용하여 결측값이 있는지 확인하고, 잠재적인 문제가 있는지 살펴볼 수 있다.**

In [5]:
df.describe()

/Users/kimminyoung/opt/anaconda3/envs/danal/lib/python3.8/site-packages/numpy/lib/function_base.py:4573: RuntimeWarning: invalid value encountered in subtract
  diff_b_a = subtract(b, a)


PRCP        SNOW   SNWD         TMAX        TMIN        TOBS  \
count  765.000000  577.000000  577.0   765.000000  765.000000  398.000000   
mean     5.360392    4.202773    NaN  2649.175294  -15.914379    8.632161   
std     10.002138   25.086077    NaN  2744.156281   24.242849    9.815054   
min      0.000000    0.000000   -inf   -11.700000  -40.000000  -16.100000   
25%      0.000000    0.000000    NaN    13.300000  -40.000000    0.150000   
50%      0.000000    0.000000    NaN    32.800000  -11.100000    8.300000   
75%      5.800000    0.000000    NaN  5505.000000    6.700000   18.300000   
max     61.700000  229.000000    inf  5505.000000   23.900000   26.100000   

            WESF  
count  11.000000  
mean   16.290909  
std     9.489832  
min     1.800000  
25%     8.600000  
50%    19.300000  
75%    24.900000  
max    28.700000

In [6]:
df.SNWD.unique()

array([-inf,  inf,  nan])

In [7]:
df.SNWD.value_counts()

-inf    553
 inf     24
Name: SNWD, dtype: int64

->
* SNWD 열은 쓸모없는 것으로 보인다. NaN값과 inf, -inf로 이루어져 있다. 
* TMAX 열은 신뢰할 수 없다. 너무 높은 온도 5505도가 있을 수 없다.

**3. `info()` 메서드를 사용하면 결측값이 있는지 알 수 있으며 열의 데이터 유형이 예상한 유형인지 확인할 수 있다.**

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 765 entries, 0 to 764
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   date               765 non-null    object 
 1   station            765 non-null    object 
 2   PRCP               765 non-null    float64
 3   SNOW               577 non-null    float64
 4   SNWD               577 non-null    float64
 5   TMAX               765 non-null    float64
 6   TMIN               765 non-null    float64
 7   TOBS               398 non-null    float64
 8   WESF               11 non-null     float64
 9   inclement_weather  408 non-null    object 
dtypes: float64(7), object(3)
memory usage: 59.9+ KB


info 결과를 보면총 765개의 row가 있지만, 5개의 열은 765보다 작다.

**4. `isna()`/`isnull()` 메서드를 사용해 null값이 들어간 행이 모두 포함되도록 데이터 프레임 생성.**

In [9]:
contain_nulls = df[
    df.SNOW.isna() | df.SNWD.isna() | df.TOBS.isna()
    | df.WESF.isna() | df.inclement_weather.isna()
]
contain_nulls.shape[0]

765

In [10]:
contain_nulls.head(10)

date            station  PRCP   SNOW  SNWD    TMAX  TMIN  \
0  2018-01-01T00:00:00                  ?   0.0    0.0  -inf  5505.0 -40.0   
1  2018-01-01T00:00:00                  ?   0.0    0.0  -inf  5505.0 -40.0   
2  2018-01-01T00:00:00                  ?   0.0    0.0  -inf  5505.0 -40.0   
3  2018-01-02T00:00:00  GHCND:USC00280907   0.0    0.0  -inf    -8.3 -16.1   
4  2018-01-03T00:00:00  GHCND:USC00280907   0.0    0.0  -inf    -4.4 -13.9   
5  2018-01-03T00:00:00  GHCND:USC00280907   0.0    0.0  -inf    -4.4 -13.9   
6  2018-01-03T00:00:00  GHCND:USC00280907   0.0    0.0  -inf    -4.4 -13.9   
7  2018-01-04T00:00:00                  ?  20.6  229.0   inf  5505.0 -40.0   
8  2018-01-04T00:00:00                  ?  20.6  229.0   inf  5505.0 -40.0   
9  2018-01-05T00:00:00                  ?   0.3    NaN   NaN  5505.0 -40.0   

   TOBS  WESF inclement_weather  
0   NaN   NaN               NaN  
1   NaN   NaN               NaN  
2   NaN   NaN               NaN  
3 -12.2   NaN             False  
4 -13.3   NaN             False  
5 -13.3   NaN             False  
6 -13.3   NaN             False  
7   NaN  19.3              True  
8   NaN  19.3              True  
9   NaN   NaN               NaN

`NaN` 값을 아래와 같은 방식으로 확인할 수 없다는 것을 주의해야한다.

데이터의 `NaN`값은 어떠한 값과도 같지 않기 때문에 **isna(), isnull()** 메서드를 사용해야한다.

In [11]:
df[df.inclement_weather == 'NaN'].shape[0]

0

In [12]:
import numpy as np
df[df.inclement_weather == np.nan].shape[0]

0

In [13]:
df[df.inclement_weather.isna()].shape[0]

357

`-inf`/`inf`는 실제로 `-np.inf`/`np.inf`이기 때문에,

`-np.inf`/`np.inf`로 음의무한대와 양의 무한대의 개수를 세어야 한다.

In [14]:
df[df.SNWD.isin([-np.inf, np.inf])].shape[0]

577

In [15]:
# 딕셔너리 컴프리헨션 사용
def get_inf_count(df):
    """데이터 프레임에서 컬럼 별로 inf와 -inf를 세는 함수"""
    return {
        col: df[df[col].isin([np.inf, -np.inf])].shape[0] for col in df.columns
    }

get_inf_count(df)

{'date': 0,
 'station': 0,
 'PRCP': 0,
 'SNOW': 0,
 'SNWD': 577,
 'TMAX': 0,
 'TMIN': 0,
 'TOBS': 0,
 'WESF': 0,
 'inclement_weather': 0}

적설량의 무한대 값을 어떻게 처리할 것인지 결정하기 전에 우리는 강설량에 대한 요약통계를 살펴봐야 한다. 강설량은 적설량을 결정하는데에 많은 비중을 차지한다.

inf와 -inf 각각으로 시리즈의 요약통계를 만들어 DataFrame 구조에 넣는다.

In [16]:
pd.DataFrame({
    'np.inf Snow Depth': df[df.SNWD == np.inf].SNOW.describe(),
    '-np.inf Snow Depth': df[df.SNWD == -np.inf].SNOW.describe()
}).T #데이터를 쉽게 보기 위해 전치

count        mean        std   min   25%    50%    75%  \
np.inf Snow Depth    24.0  101.041667  74.498018  13.0  25.0  120.5  152.0   
-np.inf Snow Depth  553.0    0.000000   0.000000   0.0   0.0    0.0    0.0   

                      max  
np.inf Snow Depth   229.0  
-np.inf Snow Depth    0.0

적설량은 눈이 내리지 않았을 때, 음의 무한대로 기록됐다.

고정된 날짜 범위에 대해 작업해야 하는경우 눈이 오지 않았다면 적설량은 0이나 NaN으로 처리하는 것이 좋겠다.

하지만, 양의 무한대는 어떻게 봐야할지 모르겠다. 그대로 놔두거나 사용하지 않는 것이 좋겠다.

**5. `date` and `station` 컬럼에 대해 요약통계를 생성해보면,**

* station 필드에는 총 2개의 값이 있는데 top값으로 나와있는 GHCND:USC00280907	와 앞에서 살펴봤듯이 ?인 것으로 알 수 있다.

* date필드에서는 freq값을 보면 어떤 날짜가 8번이나 들어가 있는 것으로 보인다. 우리는 unique값을 통해 날짜는 총 324날이 있는 것을 알 수 있다. 관측소 ID는 2개의 값이 들어가 있으므로 매일 2개의 항목이 있다고 가정해보면 648개의 데이터만 있어도 된다.

* 총 데이터의 개수는 765개 이므로 우리는 아직 무언가를 놓치고 있다.

In [17]:
df.describe(include='object') #문자열 열에 대한 통계적 요약 정보

date            station inclement_weather
count                   765                765               408
unique                  324                  2                 2
top     2018-07-05T00:00:00  GHCND:USC00280907             False
freq                      8                398               384

`duplicated()` 메서드의 결과를 부울마스크로 사용하면 중복된 행을 찾을 수 있다.

In [18]:
df[df.duplicated()].shape[0]

284

`keep` 인수의 default 값은 `'first'` 로 중복되는 행들 중 첫번째 행 외의 추가 행만 가져온다.

아래에서는 keep = False 로 설정해 한 번 이상 나타나는 모든 행을 가져온다.

In [19]:
df[df.duplicated(keep=False)].shape[0]

482

사용할 컬럼들만 따로 볼 수도 있다.

In [20]:
df[df.duplicated(['date', 'station'])].shape[0]

284

In [17]:
df[df.duplicated()].head()

date            station  PRCP   SNOW  SNWD    TMAX  TMIN  \
1  2018-01-01T00:00:00                  ?   0.0    0.0  -inf  5505.0 -40.0   
2  2018-01-01T00:00:00                  ?   0.0    0.0  -inf  5505.0 -40.0   
5  2018-01-03T00:00:00  GHCND:USC00280907   0.0    0.0  -inf    -4.4 -13.9   
6  2018-01-03T00:00:00  GHCND:USC00280907   0.0    0.0  -inf    -4.4 -13.9   
8  2018-01-04T00:00:00                  ?  20.6  229.0   inf  5505.0 -40.0   

   TOBS  WESF inclement_weather  
1   NaN   NaN               NaN  
2   NaN   NaN               NaN  
5 -13.3   NaN             False  
6 -13.3   NaN             False  
8   NaN  19.3              True

->
* 위의 5개의 행만 봤을 때, 어떤 행은 적어도 세 번 반복된다는 것을 알 수 있다. (첫 행은 포함되지 않으므로)

* 이는 1행과 2행의 데이터에서 다른 대응값을 갖는다는 것을 뜻한다.

## Mitigating Issues
### 문제 완화하기

### Handling duplicated data
Since we know we have NY weather data and noticed we only had two entries for `station`, we may decide to drop the `station` column because we are only interested in the weather data. However, when dealing with duplicate data, we need to think of the ramifications of removing it. Notice we only have data for the `WESF` column when the station is `?`:

In [18]:
df[df.WESF.notna()].station.unique()

array(['?'], dtype=object)

If we determine it won't impact our analysis, we can use `drop_duplicates()` to remove them:

In [19]:
# 1. make the date a datetime
df.date = pd.to_datetime(df.date)

# 2. save this information for later
station_qm_wesf = df[df.station == '?'].drop_duplicates('date').set_index('date').WESF

# 3. sort ? to the bottom
df.sort_values('station', ascending=False, inplace=True)

# 4. drop duplicates based on the date column keeping the first occurrence 
# which will be the valid station if it has data
df_deduped = df.drop_duplicates('date')

# 5. remove the station column because we are done with it
df_deduped = df_deduped.drop(columns='station').set_index('date').sort_index()

# 6. take valid station's WESF and fall back on station ? if it is null
df_deduped = df_deduped.assign(
    WESF=lambda x: x.WESF.combine_first(station_qm_wesf)
)

df_deduped.shape

(324, 8)

Here we used the `combine_first()` method to coalesce the values to the first non-null entry; this means that if we had data from both stations, we would first take the value provided by the named station and if (and only if) that station was null would we take the value from the station named `?`. The following table contains some examples of how this would play out:

| station GHCND:USC00280907 | station ? | result of `combine_first()` |
| :---: | :---: | :---: |
| 1 | 17 | 1 |
| 1 | `NaN` | 1 |
| `NaN` | 17 | 17 |
| `NaN` | `NaN` | `NaN` |

Check out the 4th row&mdash;we have `WESF` in the correct spot thanks to the index:

In [20]:
df_deduped.head()

PRCP   SNOW  SNWD    TMAX  TMIN  TOBS  WESF inclement_weather
date                                                                     
2018-01-01   0.0    0.0  -inf  5505.0 -40.0   NaN   NaN               NaN
2018-01-02   0.0    0.0  -inf    -8.3 -16.1 -12.2   NaN             False
2018-01-03   0.0    0.0  -inf    -4.4 -13.9 -13.3   NaN             False
2018-01-04  20.6  229.0   inf  5505.0 -40.0   NaN  19.3              True
2018-01-05  14.2  127.0   inf    -4.4 -13.9 -13.9   NaN              True

### Dealing with nulls
We could drop nulls, replace them with some arbitrary value, or impute them using the surrounding data. Each of these options may have ramifications, so we must choose wisely.

We can use `dropna()` to drop rows where any column has a null value. The default options leave us hardly any data:

In [21]:
df_deduped.dropna().shape

(4, 8)

If we pass `how='all'`, we can choose to only drop rows where everything is null, but this removes nothing:

In [22]:
df_deduped.dropna(how='all').shape

(324, 8)

We can use just a subset of columns to determine what to drop with the `subset` argument:

In [23]:
df_deduped.dropna(
    how='all', subset=['inclement_weather', 'SNOW', 'SNWD']
).shape

(293, 8)

This can also be performed along columns, and we can also require a certain number of null values before we drop the data:

In [24]:
df_deduped.dropna(axis='columns', thresh=df_deduped.shape[0] * .75).columns

Index(['PRCP', 'SNOW', 'SNWD', 'TMAX', 'TMIN', 'TOBS', 'inclement_weather'], dtype='object')

We can choose to fill in the null values instead with `fillna()`:

In [25]:
df_deduped.loc[:,'WESF'].fillna(0, inplace=True)
df_deduped.head()

PRCP   SNOW  SNWD    TMAX  TMIN  TOBS  WESF inclement_weather
date                                                                     
2018-01-01   0.0    0.0  -inf  5505.0 -40.0   NaN   0.0               NaN
2018-01-02   0.0    0.0  -inf    -8.3 -16.1 -12.2   0.0             False
2018-01-03   0.0    0.0  -inf    -4.4 -13.9 -13.3   0.0             False
2018-01-04  20.6  229.0   inf  5505.0 -40.0   NaN  19.3              True
2018-01-05  14.2  127.0   inf    -4.4 -13.9 -13.9   0.0              True

At this point we have done everything we can without distorting the data. We know that we are missing dates, but if we reindex, we don't know how to fill in the `NaN` data. With the weather data, we can't assume because it snowed one day that it will snow the next or that the temperature will be the same. For this reason, note that the next few examples are just for illustrative purposes only—just because we can do something doesn't mean we should.

That being said, let's try to address some of remaining issues with the temperature data. We know that when `TMAX` is the temperature of the Sun, it must be because there was no measured value, so let's replace it with `NaN`. We will also do so for `TMIN` which currently uses -40°C for its placeholder when we know that the coldest temperature ever recorded in NYC was -15°F (-26.1°C) on February 9, 1934:

In [26]:
df_deduped = df_deduped.assign(
    TMAX=lambda x: x.TMAX.replace(5505, np.nan),
    TMIN=lambda x: x.TMIN.replace(-40, np.nan),
)

We will also make an assumption that the temperature won't change drastically day-to-day. Note that this is actually a big assumption, but it will allow us to understand how `fillna()` works when we provide a strategy through the `method` parameter. The `fillna()` method gives us 2 options for the `method` parameter:
- `'ffill'` to forward-fill
- `'bfill'` to back-fill

*Note that `'nearest'` is missing because we are not reindexing.*

Here, we will use `'ffill'` to show how this works:

In [27]:
df_deduped.assign(
    TMAX=lambda x: x.TMAX.fillna(method='ffill'),
    TMIN=lambda x: x.TMIN.fillna(method='ffill')
).head()

PRCP   SNOW  SNWD  TMAX  TMIN  TOBS  WESF inclement_weather
date                                                                   
2018-01-01   0.0    0.0  -inf   NaN   NaN   NaN   0.0               NaN
2018-01-02   0.0    0.0  -inf  -8.3 -16.1 -12.2   0.0             False
2018-01-03   0.0    0.0  -inf  -4.4 -13.9 -13.3   0.0             False
2018-01-04  20.6  229.0   inf  -4.4 -13.9   NaN  19.3              True
2018-01-05  14.2  127.0   inf  -4.4 -13.9 -13.9   0.0              True

We can use `np.nan_to_num()` to turn `np.nan` into 0 and `-np.inf`/`np.inf` into large negative or positive finite numbers:

In [28]:
df_deduped.assign(
    SNWD=lambda x: np.nan_to_num(x.SNWD)
).head()

PRCP   SNOW           SNWD  TMAX  TMIN  TOBS  WESF  \
date                                                             
2018-01-01   0.0    0.0 -1.797693e+308   NaN   NaN   NaN   0.0   
2018-01-02   0.0    0.0 -1.797693e+308  -8.3 -16.1 -12.2   0.0   
2018-01-03   0.0    0.0 -1.797693e+308  -4.4 -13.9 -13.3   0.0   
2018-01-04  20.6  229.0  1.797693e+308   NaN   NaN   NaN  19.3   
2018-01-05  14.2  127.0  1.797693e+308  -4.4 -13.9 -13.9   0.0   

           inclement_weather  
date                          
2018-01-01               NaN  
2018-01-02             False  
2018-01-03             False  
2018-01-04              True  
2018-01-05              True

Depending on the data we are working with, we can use the `clip()` method as an alternative to `np.nan_to_num()`. The `clip()` method makes it possible to cap values at a specific minimum and/or maximum threshold. Since `SNWD` can't be negative, let's use `clip()` to enforce a lower bound of zero. To show how the upper bound works, let's use the value of `SNOW`:

In [29]:
df_deduped.assign(
    SNWD=lambda x: x.SNWD.clip(0, x.SNOW)
).head()

PRCP   SNOW   SNWD  TMAX  TMIN  TOBS  WESF inclement_weather
date                                                                    
2018-01-01   0.0    0.0    0.0   NaN   NaN   NaN   0.0               NaN
2018-01-02   0.0    0.0    0.0  -8.3 -16.1 -12.2   0.0             False
2018-01-03   0.0    0.0    0.0  -4.4 -13.9 -13.3   0.0             False
2018-01-04  20.6  229.0  229.0   NaN   NaN   NaN  19.3              True
2018-01-05  14.2  127.0  127.0  -4.4 -13.9 -13.9   0.0              True

We can couple `fillna()` with other types of calculations. Here we replace missing values of `TMAX` with the median of all `TMAX` values, `TMIN` with the median of all `TMIN` values, and `TOBS` to the average of the `TMAX` and `TMIN` values. Since we place `TOBS` last, we have access to the imputed values for `TMIN` and `TMAX` in the calculation:

In [30]:
df_deduped.assign(
    TMAX=lambda x: x.TMAX.fillna(x.TMAX.median()),
    TMIN=lambda x: x.TMIN.fillna(x.TMIN.median()),
    # average of TMAX and TMIN
    TOBS=lambda x: x.TOBS.fillna((x.TMAX + x.TMIN) / 2)
).head()

PRCP   SNOW  SNWD  TMAX  TMIN  TOBS  WESF inclement_weather
date                                                                   
2018-01-01   0.0    0.0  -inf  14.4   5.6  10.0   0.0               NaN
2018-01-02   0.0    0.0  -inf  -8.3 -16.1 -12.2   0.0             False
2018-01-03   0.0    0.0  -inf  -4.4 -13.9 -13.3   0.0             False
2018-01-04  20.6  229.0   inf  14.4   5.6  10.0  19.3              True
2018-01-05  14.2  127.0   inf  -4.4 -13.9 -13.9   0.0              True

We can also use `apply()` for running the same calculation across columns. For example, let's fill all missing values with their rolling 7-day median of their values, setting the number of periods required for the calculation to 0 to ensure we don't introduce more extra `NaN` values. Rolling calculations will be covered in chapter 4, so this is a preview:

In [31]:
df_deduped.apply(
    # rolling calculations will be covered in chapter 4, this is a rolling 7-day median
    # we set min_periods (# of periods required for calculation) to 0 so we always get a result 
    lambda x: x.fillna(x.rolling(7, min_periods=0).median())
).head(10)

PRCP   SNOW  SNWD   TMAX  TMIN   TOBS  WESF inclement_weather
date                                                                     
2018-01-01   0.0    0.0  -inf    NaN   NaN    NaN   0.0               NaN
2018-01-02   0.0    0.0  -inf  -8.30 -16.1 -12.20   0.0             False
2018-01-03   0.0    0.0  -inf  -4.40 -13.9 -13.30   0.0             False
2018-01-04  20.6  229.0   inf  -6.35 -15.0 -12.75  19.3              True
2018-01-05  14.2  127.0   inf  -4.40 -13.9 -13.90   0.0              True
2018-01-06   0.0    0.0  -inf -10.00 -15.6 -15.00   0.0             False
2018-01-07   0.0    0.0  -inf -11.70 -17.2 -16.10   0.0             False
2018-01-08   0.0    0.0  -inf  -7.80 -16.7  -8.30   0.0             False
2018-01-10   0.0    0.0  -inf   5.00  -7.8  -7.80   0.0             False
2018-01-11   0.0    0.0  -inf   4.40  -7.8   1.10   0.0             False

The last strategy we could try is interpolation with the `interpolate()` method. We specify the `method` parameter with the interpolation strategy to use. There are many options, but we will stick with the default of `'linear'`, which will treat values as evenly spaced and place missing values in the middle of existing ones. We have some missing data, so we will reindex first. Look at January 9th, which we didn't have before—the values for `TMAX`, `TMIN`, and `TOBS` are the average of values the day prior (January 8th) and the day after (January 10th):

In [32]:
df_deduped\
    .reindex(pd.date_range('2018-01-01', '2018-12-31', freq='D'))\
    .apply(lambda x: x.interpolate())\
    .head(10)

PRCP   SNOW  SNWD  TMAX   TMIN   TOBS  WESF inclement_weather
2018-01-01   0.0    0.0  -inf   NaN    NaN    NaN   0.0               NaN
2018-01-02   0.0    0.0  -inf  -8.3 -16.10 -12.20   0.0             False
2018-01-03   0.0    0.0  -inf  -4.4 -13.90 -13.30   0.0             False
2018-01-04  20.6  229.0   inf  -4.4 -13.90 -13.60  19.3              True
2018-01-05  14.2  127.0   inf  -4.4 -13.90 -13.90   0.0              True
2018-01-06   0.0    0.0  -inf -10.0 -15.60 -15.00   0.0             False
2018-01-07   0.0    0.0  -inf -11.7 -17.20 -16.10   0.0             False
2018-01-08   0.0    0.0  -inf  -7.8 -16.70  -8.30   0.0             False
2018-01-09   0.0    0.0  -inf  -1.4 -12.25  -8.05   0.0               NaN
2018-01-10   0.0    0.0  -inf   5.0  -7.80  -7.80   0.0             False

<hr>

<div style="overflow: hidden; margin-bottom: 10px;">
    <div style="float: left;">
         <a href="./4-reshaping_data.ipynb">
            <button>&#8592; Previous Notebook</button>
        </a>
    </div>
    <div style="float: right;">
        <a href="../../solutions/ch_03/solutions.ipynb">
            <button>Solutions</button>
        </a>
        <a href="../ch_04/1-querying_and_merging.ipynb">
            <button>Chapter 4 &#8594;</button>
        </a>
    </div>
</div>
<hr>